Bluetooth: Controller: nRF54L: Fix to improve decryption speed

cvinayak opened the pull request:
Fix to improve decryption speed. Decryption starts after payload reception, hence use fastest mode to decrypt PDUs.

cvinayak committed:
Bluetooth: Controller: nRF54L: Fix to improve decryption speed

Fix to improve decryption speed. Decryption starts after
payload reception, hence use fastest mode to decrypt PDUs.
nRF54Lx only supports decryption after payload reception,
this implementation was already there in the code.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Revert relaxed radio packet assignment deadline

Reverts relaxed radio packet assignment deadline as it is
too risky and can cause invalid bits be transmitted and/or
cause MIC failures.

Reverts commit 4dbfb22a7ea5 ("Bluetooth: Controller: Relax
radio packet pointer assignment deadline"), and
commit 230df7799342 ("Bluetooth: Controller: Relax radio
packet pointer assignment deadline").

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_peripheral_iso.c:
```diff
@@ -994,6 +1023,13 @@ static void isr_tx(void *param)
 
 	radio_tmr_hcto_configure_abs(hcto);
 
+#if defined(CONFIG_BT_CTLR_PROFILE_ISR) || \
+	defined(HAL_RADIO_GPIO_HAVE_PA_PIN)
+	radio_tmr_end_capture();
+#endif /* CONFIG_BT_CTLR_PROFILE_ISR ||
+	* HAL_RADIO_GPIO_HAVE_PA_PIN
+	*/
+
```
The additional newlines here are not needed

cvinayak replied:
It is a single blank line before the next line with `#if` (blank line 1032 between 1031 and 1033), right?

Thalley replied:
I meant that instead of 
```c
#if defined(CONFIG_BT_CTLR_PROFILE_ISR) || \
	defined(HAL_RADIO_GPIO_HAVE_PA_PIN)
	radio_tmr_end_capture();
#endif /* CONFIG_BT_CTLR_PROFILE_ISR ||
	* HAL_RADIO_GPIO_HAVE_PA_PIN
	*/
```
This could simply be

```c
#if defined(CONFIG_BT_CTLR_PROFILE_ISR) || defined(HAL_RADIO_GPIO_HAVE_PA_PIN)
	radio_tmr_end_capture();
#endif /* CONFIG_BT_CTLR_PROFILE_ISR || HAL_RADIO_GPIO_HAVE_PA_PIN */
```

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -2104,10 +2104,13 @@ static void *radio_ccm_ext_rx_pkt_set(struct ccm *cnf, uint8_t phy, uint8_t pdu_
+	/* NOTE: Use fast decryption as rx data is decrypt after payload is received, compared to
+	 *       decrypting in parallel with radio reception of address in nRF51/nRF52/nRF53.
+	 */
+	mode = (CCM_MODE_MODE_FastDecryption << CCM_MODE_MODE_Pos) &
```
`CCM_MODE_MODE_FastDecryption` only exists for `CONFIG_SOC_COMPATIBLE_NRF54LX`?

cvinayak replied:
yes and later series

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -2117,20 +2120,26 @@ static void *radio_ccm_ext_rx_pkt_set(struct ccm *cnf, uint8_t phy, uint8_t pdu_
 		CCM_MODE_MACLEN_Msk;
 
 #elif !defined(CONFIG_SOC_SERIES_NRF51X)
```
Consider doing 

```c
#if defined(CONFIG_SOC_COMPATIBLE_NRF54LX)
#elif defined(CONFIG_SOC_SERIES_NRF51X)
#else /* !CONFIG_SOC_COMPATIBLE_NRF54LX && !CONFIG_SOC_SERIES_NRF51X */
#endif
```

instead. Having the `else` part in the middle makes it harder to follow.

cvinayak replied:
`!defined(CONFIG_SOC_SERIES_NRF51X)`  applies to nRF52 and nRF53. There is no `CCM_MODE_LENGTH_Extended`  in nRF51.

Your suggestion will be then:
```
#if defined(CONFIG_SOC_COMPATIBLE_NRF54LX)
/* some code here */
#elif defined(CONFIG_SOC_SERIES_NRF51X)
/* no code here */
#else /* !CONFIG_SOC_COMPATIBLE_NRF54LX && !CONFIG_SOC_SERIES_NRF51X */
/* some code here */
#endif
```

Thalley replied:
> Your suggestion will be then:

Yeah, so that the "else" clause is at the end (which `!defined(CONFIG_SOC_SERIES_NRF51X)` effectively is here)

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -2167,9 +2175,10 @@ static void *radio_ccm_ext_rx_pkt_set(struct ccm *cnf, uint8_t phy, uint8_t pdu_
 	case PHY_CODED:
-		mode |= (CCM_MODE_DATARATE_125Kbps <<
-			 CCM_MODE_DATARATE_Pos) &
+#if !defined(CONFIG_SOC_COMPATIBLE_NRF54LX)
+		mode |= (CCM_MODE_DATARATE_125Kbps << CCM_MODE_DATARATE_Pos) &
 			CCM_MODE_DATARATE_Msk;
```
Given how `mode` is set and modified through this function, I would consider either defining the `mode` as a macro or a function that takes the PHY as input. 

The current implementation where `mode` is assigned 3 different places and then has 7 `|=` over 80 lines is hard to follow, and likely prone to errors.

Thalley replied:
> Different SoCs have different supported bitmaps and features to be assigned

I realize that, and that was my argument to why such a function should exist - Understanding the final value of `mode` in this function is quite hard as it's assigned and updated so many places with so many different conditionals

Thalley replied:
But since this was the code before, it's not a blocker for this PR. It does however IMO reduce readability and maintainability

Thalley reviewed (commented):
The changes look functionally OK, but I think it has poor maintainability

cvinayak committed:
Bluetooth: Controller: Use CONFIG_SOC_COMPATIBLE_NRF5340_CPUNET

Use CONFIG_SOC_COMPATIBLE_NRF5340_CPUNET for nRF53 SoC
conditional compilations.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>
